include/seastar/core/metrics_api.hh

```diff
@@ -381,6 +443,109 @@ std::unique_ptr<metric_groups_def> create_metric_groups();
 
 }
 
+/*!
+ * \brief a relabel_config allows changing metrics labels dynamically
+ *
+ * The logic is similar to Prometheus configuration
+ * This is how Prometheus entry looks like:
+ *  - source_labels: [version]
+      regex: '([0-9]+\.[0-9]+)(\.?[0-9]*).*'
+      replacement: '$1$2'
+      target_label: svr
+ */
+class relabel_config {
+public:
+    enum class action {skipwhenempty, reportwhenempty, replace, keep, drop, droplabel};
+private:
+    std::vector<std::string> _source_labels;
+    std::string _target_label;
+    std::string _replacement;
+    // _regex_str used just so we can report what the original regex was
+    std::string _regex_str;
+    std::regex _regex;
+    action _action;
+    std::string _separator;
+public:
+    relabel_config(const std::vector<std::string>& source_labels = {}, const std::string& target_label = "", const std::string& replacement = "${1}",
+            const std::string& expr = "(.*)", action a = action::replace, std::string separator = ";") :
+                _source_labels(source_labels), _target_label(target_label), _replacement(replacement), _regex_str(expr), _regex(std::regex(expr)), _action(a), _separator(separator) {
```

Comment:
Let's define it as a struct instead, so it can be initialized with designated initializers. The code can be moved to the apply function.

Reply:
My only problem with struct is the regex, I can add wrapper around regex that would have cast to and from string something like
```
class myregex {
    std::string _regex_str;
    std::regex _regex;
public:
   myregex(std::string expr): _regex_str(expr), _regex(std::regex(expr){}
   myregex&  operator=(std::string expr) {
      _regex_str = expr;
      _regex = std::regex(expr);
   }
  const std::string& operator(std::string) {
     return _regex_str;
  }
};
```

So it would behave like an assignable type.

Reply:
It's better to keep it as regex. It's a well understood type, people won't be surprised by it.

Reply:
> Q: how difficult is it to have several endpoints, each with its own configuration? That allows us to have lightweight and heavyweight endpoints.
> 
> Should be easy if I understand this correctly.

The question, what do you mean by endpoint?

Different Prometheus enpoint with different point? We technically can, but don't think it's worth the effort.
The last few changes allow you to use the same endpoint with label filtering on the requester side.
For example the prometheus server would check for:
http://ip:9180/metrics?level=1
or 
http://ip:9180/metrics

And would get different metrics, it's not exactly different endpoints but it does exactly the same things